storage/redis_cluster.go

```diff
@@ -123,46 +160,69 @@ func NewRedisClusterPool(isCache bool) *rediscluster.RedisCluster {
 		timeout = time.Duration(cfg.Timeout) * time.Second
 	}
 
-	poolConf := rediscluster.PoolConfig{
-		MaxIdle:        maxIdle,
-		MaxActive:      maxActive,
-		IdleTimeout:    240 * time.Second,
-		ConnectTimeout: timeout,
-		ReadTimeout:    timeout,
-		WriteTimeout:   timeout,
-		Database:       cfg.Database,
-		Password:       cfg.Password,
-		IsCluster:      cfg.EnableCluster,
-		UseTLS:         cfg.UseSSL,
-		TLSSkipVerify:  cfg.SSLInsecureSkipVerify,
-	}
-
-	// If Redis port isn't set, use default one:
 	if cfg.Port == 0 {
 		cfg.Port = defaultRedisPort
 	}
 
-	seed_redii := []map[string]string{}
+	var seedHosts []string
 
 	for h, p := range cfg.Hosts {
-		seed_redii = append(seed_redii, map[string]string{h: p})
+		addr := h + ":" + p
+		seedHosts = append(seedHosts, addr)
 	}
-	if len(seed_redii) == 0 {
-		seed_redii = append(seed_redii, map[string]string{cfg.Host: strconv.Itoa(cfg.Port)})
+
+	if len(seedHosts) == 0 {
+		addr := cfg.Host + ":" + strconv.Itoa(cfg.Port)
+		seedHosts = append(seedHosts, addr)
 	}
 
-	cluster := rediscluster.NewRedisCluster(seed_redii, poolConf, false)
-	return &cluster
+	var client GenericRedisClient
+	var tlsConfig *tls.Config
+
+	if cfg.UseSSL {
+		tlsConfig = &tls.Config{
+			InsecureSkipVerify: cfg.SSLInsecureSkipVerify,
+		}
+	}
+
+	if cfg.EnableCluster {
+		client = redis.NewClusterClient(&redis.ClusterOptions{
+			Addrs:        seedHosts,
+			Password:     cfg.Password,
+			DialTimeout:  timeout,
+			ReadTimeout:  timeout,
+			WriteTimeout: timeout,
+			IdleTimeout:  240 * time.Second,
+			PoolSize:     poolSize,
+			TLSConfig:    tlsConfig,
+		})
+	} else {
+		client = redis.NewClient(&redis.Options{
+			Addr:         seedHosts[0],
+			Password:     cfg.Password,
+			DB:           cfg.Database,
+			DialTimeout:  timeout,
+			ReadTimeout:  timeout,
+			WriteTimeout: timeout,
+			IdleTimeout:  240 * time.Second,
+			PoolSize:     poolSize,
+			TLSConfig:    tlsConfig,
+		})
```

Comment:
Is it worth using `NewUniversalClient` which is capable of connecting to single redis server, redis cluster and also sentinel?

```
// NewUniversalClient returns a new multi client. The type of client returned depends
// on the following three conditions:
//
// 1. if a MasterName is passed a sentinel-backed FailoverClient will be returned
// 2. if the number of Addrs is two or more, a ClusterClient will be returned
// 3. otherwise, a single-node redis Client will be returned.
func NewUniversalClient(opts *UniversalOptions) UniversalClient {
```

Reply:
@asoorm Thank you. Great catch. I was not happy with interface idea.